projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/README.md

```diff
@@ -0,0 +1,96 @@
+# Image-based Facial Expression Recognition Demo
+
+This folder contains an implemented demo of image_based_facial_expression_recognition method implemented by [[1]](#1).
+The demo framework has three main features:
+- Image: recognizes facial expressions in images.
+- Video: recognizes facial expressions in videos in a frame-based approach.
+- Webcam: connects to a webcam and recognizes facial expressions of the closest face detected by a face detection algorithm.
+The demo utilizes OpenCV face detector Haar Cascade [[2]](https://ieeexplore.ieee.org/abstract/document/990517) for real-time face detection.
+
+#### Running demo
+The pretrained models on AffectNet Categorical dataset are provided by [[1]](#1) which can be found [here](https://github.com/siqueira-hc/Efficient-Facial-Feature-Learning-with-Wide-Ensemble-based-Convolutional-Neural-Networks/tree/master/model/ml/trained_models/esr_9).
+Please note that the pretrained weights cannot be used for commercial purposes!
+To recognize a facial expression in images, run the following command:
+```python
+python inference_demo.py image -i ./media/jackie.jpg -pre ./pretrained -d -s 2
+```  
+
+The argument "image" indicates that the input is an image. The location of the image is specified after -i while -pre indicates the location of pretrained model weights. -d sets the display mode to true and -s 2 sets the window size to 1440 x 900.
+You can also visualize regions in the image relevant for the classification of facial expression by adding -g as arguments:
+```python
+python inference_demo.py image -i 'image_path' -pre 'pretrained_path' -d -s 2 -g
+```  
+The argument -g generates saliency maps with the Grad-CAM algorithm.
+
+To recognize a facial expression in videos, run the following command:
+```python
+python inference_demo.py video -i 'video_path' -pre 'pretrained_path' -d -f 5 -s 2
+```
+The argument "video" indicates that the input is a video. The location of the video is specified after -i. -d sets the display mode to true, -f defines the number of frames to be processed, and -s 2 sets the window size to 1440 x 900.
+
+To recognize a facial expression in images captured from a webcam, run the following command:
+```python
+python inference_demo.py webcam -pre 'pretrained_path' -d -s 2 
```

Comment:
When I try the webcam demo, I get this:

```
Traceback (most recent call last):
  File "inference_demo.py", line 394, in <module>
    main()
  File "inference_demo.py", line 386, in main
    webcam(args.webcam_id, args.display, args.gradcam, args.output,
  File "inference_demo.py", line 210, in webcam
    fer_display = FERDisplay(screen_size=screen_size, display_graph_ensemble=(not no_plot))
  File "/home/daniel/opendr/projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/gui/fer_display.py", line 133, in __init__
    dtype=np.int)
  File "/home/daniel/opendr/venv/lib/python3.8/site-packages/numpy/__init__.py", line 284, in __getattr__
    raise AttributeError("module {!r} has no attribute "
AttributeError: module 'numpy' has no attribute 'int'
```
after installing the toolkit from this branch, just me?

Reply:
This is likely unrelated to this PR, seeing the issue elsewhere. Might not be a bug.

Reply:
When i ran `python3 inference_demo.py webcam -d` i got the following error:

```
Error raised during video mode.
Traceback (most recent call last):
  File "inference_demo.py", line 394, in <module>
    main()
  File "inference_demo.py", line 386, in main
    webcam(args.webcam_id, args.display, args.gradcam, args.output,
  File "inference_demo.py", line 236, in webcam
    raise e
  File "inference_demo.py", line 222, in webcam
    fer = None if (img is None) else recognize_facial_expression(img, device, gradcam,
  File "inference_demo.py", line 161, in recognize_facial_expression
    face_coordinates = detect_face(image)
  File "inference_demo.py", line 69, in detect_face
    faces = _FACE_DETECTOR_HAAR_CASCADE.detectMultiScale(greyscale_image, scaleFactor=1.2, minNeighbors=9,
cv2.error: OpenCV(4.5.1) /tmp/pip-req-build-ms668fyv/opencv/modules/objdetect/src/cascadedetect.cpp:1689: error: (-215:Assertion failed) !empty() in function 'detectMultiScale'
```

@negarhdr is it something wrong on my end, got any ideas? :)

Reply:
It should now download the pretrained weights automatically from the FTP server.

Reply:
> When i ran `python3 inference_demo.py webcam -d` i got the following error:
> 
> ```
> Error raised during video mode.
> Traceback (most recent call last):
>   File "inference_demo.py", line 394, in <module>
>     main()
>   File "inference_demo.py", line 386, in main
>     webcam(args.webcam_id, args.display, args.gradcam, args.output,
>   File "inference_demo.py", line 236, in webcam
>     raise e
>   File "inference_demo.py", line 222, in webcam
>     fer = None if (img is None) else recognize_facial_expression(img, device, gradcam,
>   File "inference_demo.py", line 161, in recognize_facial_expression
>     face_coordinates = detect_face(image)
>   File "inference_demo.py", line 69, in detect_face
>     faces = _FACE_DETECTOR_HAAR_CASCADE.detectMultiScale(greyscale_image, scaleFactor=1.2, minNeighbors=9,
> cv2.error: OpenCV(4.5.1) /tmp/pip-req-build-ms668fyv/opencv/modules/objdetect/src/cascadedetect.cpp:1689: error: (-215:Assertion failed) !empty() in function 'detectMultiScale'
> ```
> 
> @negarhdr is it something wrong on my end, got any ideas? :)

Hi, 
apparently it couldn't read the face_detector 'xml' file. I updated the path, and it should work now. Please let me know if there is the same problem in the ROS node as well.